### ISSUE TITLE
Where to place IN layer?

### ISSUE THREAD

**Issue by luowy1001** [recong/Boundless-in-Pytorch#1]
Thanks for your great work！ 
However, where to place IN layer in the generator? I found that you place IN layer after the activation layer. 
```
        x = self.conv2d(input)
        mask = self.mask_conv2d(input)
        x = elu(x) * self.gated(mask)
        x = instance_norm(x)
```
But we employ Conv-IN-ReLU in general. Have you considered this issue？ Thx!
@recong

**Comment by recong** [comment]
Thanks for your interest.
When I implemented this, I used this implementation (https://github.com/avalonstrel/GatedConvolution_pytorch/blob/master/models/networks.py) to Gated convolution. In this repository, they implemented GatedConv2dWithActivation as follows;
```
    def forward(self, input):
        x = self.conv2d(input)
        mask = self.mask_conv2d(input)
        if self.activation is not None:
            x = self.activation(x) * self.gated(mask)
        else:
            x = x * self.gated(mask)
        if self.batch_norm:
            return self.batch_norm2d(x)
        else:
            return x
```
In the original paper, they said they used  an ELU function, so I chose self.activation=ELU. And they said ```We additionally added instance normalization [39] after every generator layer besides the output layer``` in the paper, I removed batch normalization and added instance normalization.

I think my GatedConvolution function is  correct, but I don't know Conv-IN-ReLU which you said. If you want to know more about GatedConvolution, I recommend the original paper(https://arxiv.org/pdf/1806.03589.pdf).

**Comment by luowy1001** [comment]
Thanks for your reply!
Original Deepfilllv2 (GateConv) implemented by tensorflow [https://github.com/JiahuiYu/generative_inpainting](url). In original implement, they don not use norm layer:
```
def gen_conv(x, cnum, ksize, stride=1, rate=1, name='conv',
             padding='SAME', activation=tf.nn.elu, training=True):
    assert padding in ['SYMMETRIC', 'SAME', 'REFELECT']
    if padding == 'SYMMETRIC' or padding == 'REFELECT':
        p = int(rate*(ksize-1)/2)
        x = tf.pad(x, [[0,0], [p, p], [p, p], [0,0]], mode=padding)
        padding = 'VALID'
    x = tf.layers.conv2d(
        x, cnum, ksize, stride, dilation_rate=rate,
        activation=None, padding=padding, name=name)
    if cnum == 3 or activation is None:
        # conv for output
        return x
    x, y = tf.split(x, 2, 3)
    x = activation(x)
    y = tf.nn.sigmoid(y)
    x = x * y
    return x
```
`Conv-IN-ReLU` means the sequence of operations: First norm layer, then elu layer:
```
        x = self.conv2d(input)
        mask = self.mask_conv2d(input)
        x = instance_norm(x) * self.gated(mask)
        x = elu(x)
```
In general, pepple use this sequence, for instance, resnet.
Thx!

**Comment by recong** [comment]
I understood. I think you shouldn't place IN layer before the activation layer.
Because self.gated consists of torch.nn.Conv2d and it isn't normalized.
```
        x = self.conv2d(input)
        mask = self.mask_conv2d(input)
        x = instance_norm(x) * self.gated(mask)
        x = elu(x)
```
Therefore, instance_norm(x) is normalized, but self.gated(mask) is not normalized in this code. Consequently, the output isn't normalized and this would harm the network.

However, this is not an 'issue' of my code. I think you shouldn't post it here.

**Comment by luowy1001** [comment]
You are right! Thank you!
